src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveTestsWithVariousTypes.cs

```diff
@@ -0,0 +1,234 @@
+
+                // Type comparisons
+                for (int i = 0; i < types.Length; i++)
+                {
+                    Type sourceType = types[i];
+                    Type typeFromDisk = typesFromDisk[i];
+
+                    Assert.True(typeFromDisk.IsInterface);
+                    AssertTypeProperties(sourceType, typeFromDisk);
+                    AssertMethods(sourceType.GetMethods(), typeFromDisk.GetMethods());
+                    AssertFields(sourceType.GetFields(), typeFromDisk.GetFields());
+                }
+            }
+        }
+
+        private static void AssertMethods(MethodInfo[] sourceMethods, MethodInfo[] methodsFromDisk)
+        {
+            Assert.Equal(sourceMethods.Length, methodsFromDisk.Length);
+
+            // Method comparisons
+            for (int j = 0; j < sourceMethods.Length; j++)
+            {
+                MethodInfo sourceMethod = sourceMethods[j];
+                MethodInfo methodFromDisk = methodsFromDisk[j];
+
+                Assert.Equal(sourceMethod.Name, methodFromDisk.Name);
+                Assert.Equal(sourceMethod.Attributes, methodFromDisk.Attributes);
+                Assert.Equal(sourceMethod.ReturnType.FullName, methodFromDisk.ReturnType.FullName);
+            }
+        }
+
+        private static void AssertTypeProperties(Type sourceType, Type typeFromDisk)
+        {
+            Assert.Equal(sourceType.Name, typeFromDisk.Name);
+            Assert.Equal(sourceType.Namespace, typeFromDisk.Namespace);
+            Assert.Equal(sourceType.Attributes, typeFromDisk.Attributes);
+        }
+
+        public static IEnumerable<object[]> VariousStructsTestData()
+        {
+            yield return new object[] { new Type[] { typeof(EmptyStruct) } };
+            yield return new object[] { new Type[] { typeof(StructWithField) } };
+            yield return new object[] { new Type[] { typeof(EmptyStruct), typeof(StructWithField) } };
+            yield return new object[] { new Type[] { typeof(StructWithField), typeof(EmptyStruct) } };
+        }
+
+        [Theory]
+        [MemberData(nameof(VariousStructsTestData))]
+        public void WriteVariousStructsToStream(Type[] types)
+        {
+            using var stream = new MemoryStream();
+            AssemblyTools.WriteAssemblyToStream(s_assemblyName, types, stream);
+
+            Assembly assemblyFromDisk = AssemblyTools.LoadAssemblyFromStream(stream);
+
+            Assert.Equal(s_assemblyName.Name, assemblyFromDisk.GetName().Name);
+
+            Module moduleFromDisk = assemblyFromDisk.Modules.First();
+            Type[] typesFromDisk = moduleFromDisk.GetTypes();
+
+            Assert.Equal(types.Length, typesFromDisk.Length);
+
+            for (int i = 0; i < types.Length; i++)
+            {
+                Type sourceType = types[i];
+                Type typeFromDisk = typesFromDisk[i];
+
+                Assert.True(sourceType.IsValueType);
```

Comment:
This should be part of `AssertTypeProperties` as

`Assert.Equal(sourceType.IsValueType, typeFromDisk.IsValueType)`.

Similarly, the interface check should be part of `AssertTypeProperties` as well: `Assert.Equal(sourceType.IsInterface, typeFromDisk.IsInterface)`.

Once you do that, the whole type comparison can be shared between the different tests.

Reply:
Done, look well organized, thanks